from biobase.constants import amino_acid, nucleic_acid, rna_seq

# amino acid constants
ONE_LETTER_CODES = amino_acid.ONE_LETTER_CODES
ONE_LETTER_CODES_EXT = amino_acid.ONE_LETTER_CODES_EXT
THREE_LETTER_CODES = amino_acid.THREE_LETTER_CODES
THREE_LETTER_CODES_EXT = amino_acid.THREE_LETTER_CODES_EXT
AMINO_ACID_NAMES = amino_acid.AMINO_ACID_NAMES
AMINO_ACID_NAMES_EXT = amino_acid.AMINO_ACID_NAMES_EXT
MONO_MASS = amino_acid.MONO_MASS
MONO_MASS_EXT = amino_acid.MONO_MASS_EXT
CODONS = amino_acid.CODONS
CODONS_PER_AA = amino_acid.CODONS_PER_AA
CODON_TABLE = amino_acid.CODON_TABLE

# nucleic acid constants
DNA = nucleic_acid.DNA
RNA = nucleic_acid.RNA
NUCLEOTIDES = nucleic_acid.NUCLEOTIDES
NUCLEOTIDE_NAMES = nucleic_acid.NUCLEOTIDE_NAMES
MOLECULAR_WEIGHT = nucleic_acid.MOLECULAR_WEIGHT
DNA_COMPLEMENTS = nucleic_acid.DNA_COMPLEMENTS
RNA_COMPLEMENTS = nucleic_acid.RNA_COMPLEMENTS
IUPAC_NUCLEOTIDES = nucleic_acid.IUPAC_NUCLEOTIDES
SEQUENCE_MOTIFS = nucleic_acid.SEQUENCE_MOTIFS
PURINE_BASES = nucleic_acid.PURINE_BASES
PYRIMIDINE_BASES = nucleic_acid.PYRIMIDINE_BASES
WEAK_PAIRS = nucleic_acid.WEAK_PAIRS
STRONG_PAIRS = nucleic_acid.STRONG_PAIRS
START_CODONS = nucleic_acid.START_CODONS
STOP_CODONS = nucleic_acid.STOP_CODONS
RESTRICTION_SITES = nucleic_acid.RESTRICTION_SITES
RESTRICTION_SITES_REV = nucleic_acid.RESTRICTION_SITES_REV
RESTRICTION_SITES_CUTTING = nucleic_acid.RESTRICTION_SITES_CUTTING
RESTRICTION_SITES_CUTTING_REV = nucleic_acid.RESTRICTION_SITES_CUTTING_REV

# rna seq
NORMALIZATION_METHODS = rna_seq.NORMALIZATION_METHODS
