"""Utilities for making plots with your data.

BNF
---

.. highlight:: none

.. literalinclude:: ../../doc/bnf/plotting.bnf

Parameter notes
---------------

``pisa-section``
''''''''''''''''

You can specify either a file of PISA data or a custom array.
If you give a file, you need to be very careful about coordinates!
See the documentation for loadPisa for the coordinate convention used
to make PISA plots.

``h5-name``
    The name of a PISA hdf5 file, generated by
    :py:mod:`interpretPisa<bpreveal.interpretPisa>`. Cannot be used if
    ``values`` is present.

``values``
    Instead of giving a PISA file, give an array of values to plot. These
    should already be sheared and in the form generated by
    :py:func:`loadPisa<bpreveal.utils.loadPisa>`. It will be an array of
    shape (num-regions, num-regions). Cannot be used if ``h5-name`` is given.
    Note that the cropping (in ``coordinates``) is applied *to* this array, so
    it should represent the data for the whole window and then the plotting
    functions will crop in appropriately.

``color-map``
    Either ``"clip"``, ``"noclip"``, or a ``matplotlib.colors.Colormap``.
    ``"clip"`` maps to :py:data:`pisaClip<bpreveal.colors.pisaClip`, and
    ``"noclip"`` maps to :py:data:`pisaClip<bpreveal.colors.pisaNoClip`.
    Default: ``"clip"``.

``rasterize``
    Either ``True`` or ``False``. This determines if the pisa graph or plot
    should be rasterized to a bitmap before being saved. If True, then the
    saved plot will be rasterized and cannot be edited as a vector graphic.

``coordinates-section``
'''''''''''''''''''''''

``genome-fasta``
    The name of a fasta-format file containing the genome of the organism. The
    relevant sequence information will be extracted based on
    ``genome-window-start`` and ``genome-window-chrom``. Cannot be used if
    ``sequence`` is provided.

``sequence``
    Instead of looking up the sequence in a fasta, you can provide it manually.
    This should be a string of ``A``, ``C``, ``G``, and ``T``, and it must have
    the same length as the shape of the values array (or, equivalently, the
    number of regions in your PISA hdf5 file).
    Default: If no sequence is provided, then you cannot use ``show-sequence``
    in the importance or prediction sections.

``midpoint-offset``
    For making a plot, how far from the left edge of the PISA data do you want
    the middle of the plot to be? For example, if your PISA data start at
    coordinate 131,500 and you want coordinate 132,000 to be in the middle of
    the plot, you'd set ``"midpoint-offset": 500``.

``input-slice-width``, ``output-slice-width``
    How wide of a region do you want to be plotted? ``input-slice-width``
    determines the width of the x-axis. Use a smaller value to see the effect
    of fewer and fewer bases. ``output-slice-width`` determines the height of
    the y-axis. Use smaller values to zoom in on local effects.
    For PISA graphs, these parameters have a slightly different meaning.
    ``input-slice-width`` determines how wide of a window you want plotted.
    Both the top and bottom of the graph will be this wide.
    ``output-slice-width`` determines how much you want to allow for lines
    that go past the end of the figure. If ``output-slice-width > input-slice-width``
    then any lines that would go from an input on the page to an output off the page
    (or vice-versa) will still be drawn, they'll just stop when they get to the
    edge of the plot.
    I recommend keeping ``output-slice-width`` about 200 or so larger than
    ``input-slice-width`` for PISA graphs.

``genome-window-start``, ``genome-window-chrom``
    These are used to set the values for the tick labels on the plots,
    and also to extract the sequence from a fasta file if you provided
    ``genome-fasta``.

``profile-section``
'''''''''''''''''''

The profile section gives the data that should be plotted in the importance score
or prediction tracks. The format for both tracks is identical.

``bigwig-name``
    The name of the bigwig file to read from. Data will be extracted based on
    ``genome-window-start`` and ``genome-window-chrom``. Cannot be used with
    ``values``.

``values``
    An array of shape (num-regions) containing the profile values to use at
    each base, starting at ``genome-window-start``.

``show-sequence``
    If ``true``, then draw the actual letters of the DNA sequence to represent the
    profile. If ``false``, then draw a bar plot like in IGV.
    Default: ``false``.

``color``
    Color, for a profile, can be one of several things. It can be a single
    ``color-spec``, or a dictionary with one ``color-spec`` for each DNA base.
    Alternatively, it can be a list of ``color-spec`` (or dictionary with one
    ``color-spec`` for each base) with one entry for each position in the profile.
    This way, you can color each bar (or each letter in the logo) a different
    color. See the :py:mod:`colors<bpreveal.colors>` documentation
    for a description of ``color-spec``.
    Default: If ``show-sequence`` is ``true``, then the default Wong color map
    for bases. If ``show-sequence`` is ``false``, then Tol color 0.

``data-range``
    If provided, gives the axis limits for this track. This is handy when making
    animations, since otherwise the axis can jump all over the place.
    If ``None`` (the default), then the axis limits will be determined by
    the data being shown.

``annotation-section``
''''''''''''''''''''''

Annotations are used to label specific areas of the plot, typically with
things like motifs and genes.

``bed-name``
    If you scanned for motifs, then you can give a bed file here.
    Hits in that bed file will be drawn on the plot.
    Default: Do not read in a bed file.

``name-colors``
    In order to provide a consistent color to each named motif, you can
    specify which names get which colors. It is a simple dict mapping
    names (e.g., "Abf1") to colorSpecs. Whenever an entry from the bed
    file is drawn, the code will check to see if that entry's name appears
    in this dict. If it does, then that color will be used to draw its
    annotation box.
    Default: Assign colors as names are encountered.

``custom``
    A list of annotations that you provide that do not come from the bed file.
    An annotation contains the following properties:

    ``start``
        The *genomic* start coordinate of the annotation.

    ``end``
        The *genomic* end coordinate of the annotation.

    ``name``
        The name of the thing being annotated. This will be drawn
        over the box (if ``mini`` is False) or on the right-hand legend
        (if ``mini`` is True)

    ``color``
        A color-spec that defines the color of the annotation. See
        :py:mod:`colors<bpreveal.colors>` for colorSpec documentation.

    ``shape``
        An optional parameter that determines the shape that
        will be drawn, which is used to show corruptors from the GA. Options are
        ``diamond``, ``wedge``, ``indel``, ``snp``, one of the letters
        ``ACGTdǍČǦŤ``, or ``box``. ``indel`` or one of ``dǍČǦŤ`` create a wedge,
        ``snp`` or one of ``ACGT`` create a diamond. The default is ``box``.
        You can also specify ``trapezoid,up,<number>`` or ``trapezoid,down,<number>``.
        When drawn, the trapezoid will have its thin end between ``start`` and ``stop``,
        but will bulge out by ``<number>`` bases on the other side. This is useful
        if your annotated features are small and the axis for your plot covers
        a large range. I like upward-facing (i.e., narrow end on top) trapezoids
        on PISA plots.

    ``bottom`` and ``top``
        (Optional) Explicitly set the bottom and top coordinate of the annotation box.
        (This is usually not necessary, as the annotation function automatically
        adjusts the box positions if they
        overlap.)

    ``force-color-positive`` and ``force-color-negative``
        (Optional) Instead of using the normal PISA color map for this area, use these colors.
        Each is a color-spec. Values below zero will be shown as the negative color,
        and values above zero will be shown using the positive color. This is only valid
        for pisa graphs (AKA squid plots) and has no effect on PISA heatmaps.
        ``force-color-positive`` and ``force-color-negative`` override ``use-annotation-colors``.

    Default: No custom annotations.

``figure-section``
''''''''''''''''''

``bottom``, ``left``, ``width``, ``height``
    Where on the given figure should the plot be placed? Note that axis labels
    and other text can overflow this bounding box. For figure prep, you'll
    likely have to tweak these parameters.

``annotation-height``
    How tall, as a fraction of the total height allocated for annotations,
    should the drawn boxes be? Note that some extra space will be placed
    between the boxes for clarity. For large figures, you'll probably want to
    make this smaller.
    Default: 0.13

``tick-font-size``, ``label-font-size``
    For coordinate ticks and text labels, what font size should be used?
    Defaults: ``tick-font-size`` defaults to 6 and ``label-font-size`` defaults to 8.

``color-span``
    This sets the color limit for PISA values. If a clipping color map
    is used, then any values above this will show as clipped colors.

``miniature``
    Alters the layout of the graph to be a better fit to single-column display.
    Default: ``false``.

``grid-mode``, ``diagonal-mode``
    For PISA plots only (not PISA graphs), these parameters determine whether
    or not the grid lines and diagonal line should be drawn. ``grid-mode`` can
    be either ``"on"`` or ``"off"``, while ``diagonal-mode`` can be one of
    ``"on"``, ``"off"``, or ``"edge"``. ``edge`` means that the diagonal lines
    should only be drawn at the border of the PISA plot, like inward-facing
    axis ticks.
    Defaults: ``grid-mode`` defaults to ``"on"`` and
    ``diagonal-mode`` defaults to ``"edge"``.

``line-width``
    For PISA graphs only (not plots). This sets the width of the lines that
    connect the cause to effect. For large images, increasing this reduces
    Moiré interference.
    Default: 1.0

Specific parameters
'''''''''''''''''''

``min-value``
    Only applicable to PISA graphs, not plots.
    If (the absolute value of) a PISA entry between two bases
    is less than min-value, no line will be drawn at all.
    I recommend keeping this at about the 95th percentile
    of your PISA data, as otherwise an absolutely enormous number of
    splines will be drawn.
    The color bar will be drawn as white within this range.

``use-annotation-colors``
    Only applicable to PISA graphs.
    If ``true``, then any splines that originate from a base that
    overlaps with an annotation will be drawn in the color of that
    annotation instead of the normal color map. This is useful
    for showing the effects of different motifs, for example.
    If ``false``, then all splines will use the default color map
    based on the PISA value between the bases they connect.
    Note that if you specify ``force-color-positive`` and
    ``force-color-negative`` on one of your annotations, those
    colors will be used even if ``use-annotation-colors`` is ``False``.
    Default: ``false``.

Module contents
---------------

"""
from typing import Literal
from collections.abc import Iterable
import numpy as np

import matplotlib.figure
from matplotlib.axes import Axes as AXES_T

from bpreveal import logUtils
from bpreveal.internal.constants import DNA_COLOR_SPEC_T, ONEHOT_T, PRED_AR_T, ONEHOT_AR_T, \
    FONT_FAMILY, FONT_SIZE_TICKS, FONT_SIZE_LABELS, NUM_BASES, PRED_T
import bpreveal.internal.plotUtils as pu
from bpreveal import utils
from bpreveal import motifUtils
from bpreveal import schema
import bpreveal.colors as bprcolors
from bpreveal.colors import parseSpec
from numpy.typing import NDArray


def plotPisaGraph(config: dict, fig: matplotlib.figure.Figure, validate: bool = True) -> dict:
    r"""Make a graph-style PISA plot.

    :param config: The JSON (or any dictionary, really) configuration for this PISA plot.
    :param fig: The matplotlib figure that will be drawn on.
    :param validate: Should the configuration be checked? Set to False if the validator is
        whining about something that you know is a false alarm. Default: True
    :return: A dictionary containing the created Axes objects, the assigned colors for
        annotations, and the genomic coordinates is the plot.

    The structure of the returned dict is:

    .. highlight:: none

    ::

        {"axes": {"pisa": matplotlib.axes,
                  "importance": matplotlib.axes,
                  "predictions": matplotlib.axes,
                  "annotations": matplotlib.axes,
                  "cbar": matplotlib.axes,
                  "legend": matplotlib.axes}
            "name-colors": Same structure as the config dict, but with missing entries added.
            "genome-start": int,
            "genome-end": int,
            "config": config dict with data loaded.}

    .. highlight:: python

    Example::

        fig = plt.figure(figsize=(7,4))
        pisaSection = {"h5-name": f"{WORK_DIR}/shap/pisa_nanog_positive.h5"}

        coordinatesSection = {
            "genome-fasta": GENOME_FASTA,
            "midpoint-offset": 1150,
            "input-slice-width": 200,
            "output-slice-width": 500,
            "genome-window-start": windowStart,
            "genome-window-chrom": windowChrom
        }

        predictionSection = {"bigwig-name": f"{WORK_DIR}/pred/nanog_residual_positive.bw"}

        importanceSection = {
            "bigwig-name": f"{WORK_DIR}/shap/nanog_profile.bw",
            "show-sequence": True
        }

        annotationSection = {"bed-name": f"{WORK_DIR}/scan/nanog_profile.bed"}

        figureSectionGraph = {
            "left": 0.12,
            "bottom": 0.13,
            "width": 0.8,
            "height": 0.85,
            "color-span": 0.5,
        }

        graphConfig = {
            "pisa": pisaSection,
            "coordinates": coordinatesSection,
            "importance": importanceSection,
            "predictions": predictionSection,
            "annotations": annotationSection,
            "figure": figureSectionGraph,
            "min-value": 0.1,
            "use-annotation-colors": False
        }

        bpreveal.plotting.plotPisaGraph(graphConfig, fig);

    This code produces a graph that looks like this:

    .. image:: ../../doc/presentations/pisaGraph.png
        :width: 800
        :alt: Representative PISA graph.


    By turning ``miniature`` to ``true``, and setting the figure size to (3, 2), you
    instead get a plot like this:

    .. image:: ../../doc/presentations/pisaGraphMini.png
        :width: 400
        :alt: Representative miniature PISA graph.

    (Actually, this particular graph had overlapping labels that I had to remove with
    :py:func:`~deleteTick`.)
    """
    if validate:
        schema.pisaGraph.validate(config)
    cfg = pu.buildConfig(config)
    del config  # Don't accidentally use the old one.
    logUtils.debug("Starting to draw PISA graph.")
    mini = cfg["figure"]["miniature"]
    axGraph, axSeq, axProfile, axAnnot, axCbar, axLegend = pu.getPisaGraphAxes(
        fig=fig, left=cfg["figure"]["left"], bottom=cfg["figure"]["bottom"],
        width=cfg["figure"]["width"], height=cfg["figure"]["height"],
        mini=mini)

    coords = cfg["coordinates"]
    sliceStart = coords["midpoint-offset"] - coords["input-slice-width"] // 2
    sliceEnd = sliceStart + coords["input-slice-width"]
    genomeStart = coords["genome-window-start"] + sliceStart
    genomeEnd = coords["genome-window-start"] + sliceEnd

    colorBlocks = []
    if cfg["use-annotation-colors"]:
        for annot in cfg["annotations"]["custom"]:
            if "force-color-positive" not in annot:
                # We don't have a custom color scale,
                # so just use the color of the annotation for
                # the color scale.
                colorBlocks.append((annot["start"],
                                    annot["end"],
                                    annot["color"], annot["color"]))
    for annot in cfg["annotations"]["custom"]:
        if "force-color-positive" in annot:
            colorBlocks.append((annot["start"],
                                annot["end"],
                                annot["force-color-positive"],
                                annot["force-color-negative"]))

    logUtils.debug("Axes set. Drawing graph.")
    Δ = coords["output-slice-width"] - coords["input-slice-width"]
    shearMat = cfg["pisa"]["values"][sliceStart - Δ:sliceEnd + Δ,
                                     sliceStart - Δ:sliceEnd + Δ]
    pisaCax = pu.addPisaGraph(similarityMat=shearMat,
                              minValue=cfg["min-value"],
                              colorSpan=cfg["figure"]["color-span"],
                              colorBlocks=colorBlocks,
                              genomeStart=genomeStart,
                              lineWidth=cfg["figure"]["line-width"],
                              trim=Δ,
                              ax=axGraph,
                              cmap=cfg["pisa"]["color-map"],
                              rasterize=cfg["pisa"]["rasterize"])

    # Now set up the sequence/importance axis.
    logUtils.debug("Graph complete. Finishing plot.")
    pu.addHorizontalProfilePlot(
        values=cfg["importance"]["values"][sliceStart:sliceEnd],
        colors=cfg["importance"]["color"][sliceStart:sliceEnd],
        sequence=coords["sequence"][sliceStart:sliceEnd],
        genomeStartX=genomeStart, genomeEndX=genomeEnd,
        axSeq=axSeq, axGraph=axGraph,
        fontSizeTicks=cfg["figure"]["tick-font-size"],
        fontSizeAxLabel=cfg["figure"]["label-font-size"],
        showSequence=cfg["importance"]["show-sequence"],
        labelXAxis=True,
        yAxisLabel="" if mini else "Contrib.\nscore",
        dataRange=cfg["importance"]["data-range"],
        mini=mini)
    usedNames = pu.addAnnotations(
        axAnnot=axAnnot, annotations=cfg["annotations"]["custom"],
        boxHeight=cfg["figure"]["annotation-height"],
        genomeStartX=genomeStart, genomeEndX=genomeEnd,
        fontSize=cfg["figure"]["label-font-size"],
        mini=cfg["figure"]["miniature"])

    # Now, add the profiles.
    pu.addHorizontalProfilePlot(
        values=cfg["predictions"]["values"][sliceStart:sliceEnd],
        colors=cfg["predictions"]["color"][sliceStart:sliceEnd],
        sequence=coords["sequence"][sliceStart:sliceEnd],
        genomeStartX=genomeStart, genomeEndX=genomeEnd,
        axSeq=axProfile, axGraph=None,
        fontSizeTicks=cfg["figure"]["tick-font-size"],
        fontSizeAxLabel=cfg["figure"]["label-font-size"],
        showSequence=cfg["predictions"]["show-sequence"],
        labelXAxis=False,
        yAxisLabel="" if mini else "Pred.\nprofile",
        dataRange=cfg["predictions"]["data-range"],
        mini=mini)

    pu.addCbar(
        pisaCax=pisaCax, axCbar=axCbar,
        fontSizeTicks=cfg["figure"]["tick-font-size"],
        fontSizeAxLabel=cfg["figure"]["label-font-size"],
        mini=mini)

    if axLegend is not None:
        pu.addLegend(
            usedNames=usedNames,
            axLegend=axLegend,
            fontSize=cfg["figure"]["label-font-size"])

    logUtils.debug("PISA graph complete.")
    return {"axes": {"pisa": axGraph, "importance": axSeq, "predictions": axProfile,
                     "annotations": axAnnot, "cbar": axCbar, "legend": axLegend},
            "name-colors": cfg["annotations"].get("name-colors", {}),
            "genome-start": genomeStart,
            "genome-end": genomeEnd,
            "config": cfg}


def plotPisa(config: dict, fig: matplotlib.figure.Figure, validate: bool = True) -> dict:
    r"""Given the actual vectors to show, make a pretty PISA plot.

    :param config: The JSON (or any dictionary, really) configuration for this PISA plot.
    :param fig: The matplotlib figure that will be drawn on.
    :param validate: Should the configuration be checked?
    :return: A dict containing the generated axes, along with the assigned name colors and
        the coordinates that were plotted.

    The returned dict will have the following structure:

    .. highlight:: none

    ::

        {"axes": {"pisa": matplotlib.axes,
                  "importance": matplotlib.axes,
                  "predictions": matplotlib.axes,
                  "annotations": matplotlib.axes,
                  "cbar": matplotlib.axes,
                  "legend": matplotlib.axes}
            "name-colors": Same structure as the config dict, but with missing entries added.
            "genome-start": int,
            "genome-end": int,
            "config": config dict with data loaded}

    .. highlight:: python

    Example::

        fig = plt.figure(figsize=(7,4))

        pisaSection = {"h5-name": f"{WORK_DIR}/shap/pisa_nanog_positive.h5"}

        coordinatesSection = {
            "genome-fasta": GENOME_FASTA,
            "midpoint-offset": 1150,
            "input-slice-width": 200,
            "output-slice-width": 500,
            "genome-window-start": windowStart,
            "genome-window-chrom": windowChrom
        }

        predictionSection = {"bigwig-name": f"{WORK_DIR}/pred/nanog_residual_positive.bw"}

        importanceSection = {
            "bigwig-name": f"{WORK_DIR}/shap/nanog_profile.bw",
            "show-sequence": True
        }

        annotationSection = {"bed-name": f"{WORK_DIR}/scan/nanog_profile.bed"}

        figureSectionPlot = {
            "left": 0.12,
            "bottom": 0.13,
            "width": 0.8,
            "height": 0.85,
            "color-span" : 0.5,
        }

        plotConfig = {
            "pisa": pisaSection,
            "coordinates": coordinatesSection,
            "importance": importanceSection,
            "predictions": predictionSection,
            "annotations": annotationSection,
            "figure": figureSectionPlot
        }

        bpreveal.plotting.plotPisa(plotConfig, fig)

    This code produces a plot that looks like this:

    .. image:: ../../doc/presentations/pisaPlot.png
        :width: 800
        :alt: Representative PISA plot.

    By turning ``miniature`` to ``true``, and setting the figure size to (3, 2), you
    instead get a plot like this:

    .. image:: ../../doc/presentations/pisaPlotMini.png
        :width: 400
        :alt: Representative miniature PISA graph.

    (Actually, this particular plot had overlapping labels that I had to remove with
    :py:func:`~deleteTick`.)
    """
    if validate:
        schema.pisaPlot.validate(config)
    cfg = pu.buildConfig(config)
    del config  # Don't accidentally edit the old one.
    logUtils.debug("Starting to draw PISA heatmap.")
    mini = cfg["figure"]["miniature"]
    axPisa, axSeq, axProfile, axCbar, axAnnot, axLegend = pu.getPisaAxes(
        fig=fig, left=cfg["figure"]["left"], bottom=cfg["figure"]["bottom"],
        width=cfg["figure"]["width"], height=cfg["figure"]["height"],
        mini=mini)
    # PISA image plotting
    coords = cfg["coordinates"]
    sliceStartX = coords["midpoint-offset"] - coords["input-slice-width"] // 2
    sliceEndX = sliceStartX + coords["input-slice-width"]
    sliceStartY = coords["midpoint-offset"] - coords["output-slice-width"] // 2
    sliceEndY = sliceStartY + coords["output-slice-width"]
    genomeStartX = coords["genome-window-start"] + sliceStartX
    genomeEndX = coords["genome-window-start"] + sliceEndX
    shearMat = np.copy(cfg["pisa"]["values"][sliceStartY:sliceEndY, sliceStartX:sliceEndX])
    if shearMat.shape != (coords["output-slice-width"], coords["input-slice-width"]):
        logUtils.error("Your slice goes off the edge of the PISA data. Reduce your "
                       "slice widths. Plotting will continue, but axes will be incorrect. "
                       "This will trigger an error in BPReveal 6.0.0.")
    colorBlocks = []
    for annot in cfg["annotations"]["custom"]:
        colorBlocks.append((annot["start"] - coords["genome-window-start"],
                            annot["end"] - coords["genome-window-start"],
                            annot["color"]))
    logUtils.debug("Axes set. Drawing heatmap.")
    pisaCax = pu.addPisaPlot(shearMat=shearMat,
                             colorSpan=cfg["figure"]["color-span"],
                             axPisa=axPisa,
                             diagMode=cfg["figure"]["diagonal-mode"],
                             gridMode=cfg["figure"]["grid-mode"],
                             fontSizeTicks=cfg["figure"]["tick-font-size"],
                             fontSizeAxLabel=cfg["figure"]["label-font-size"],
                             genomeWindowStart=genomeStartX,
                             mini=mini,
                             cmap=cfg["pisa"]["color-map"],
                             rasterize=cfg["pisa"]["rasterize"])
    # Now set up the sequence/importance axis.
    logUtils.debug("Main plot complete. Adding horizontal profile plot.")
    pu.addHorizontalProfilePlot(
        values=cfg["importance"]["values"][sliceStartX:sliceEndX],
        colors=cfg["importance"]["color"][sliceStartX:sliceEndX],
        sequence=coords["sequence"][sliceStartX:sliceEndX],
        genomeStartX=genomeStartX, genomeEndX=genomeEndX,
        axSeq=axSeq, axGraph=axPisa,
        fontSizeTicks=cfg["figure"]["tick-font-size"],
        fontSizeAxLabel=cfg["figure"]["label-font-size"],
        showSequence=cfg["importance"]["show-sequence"],
        labelXAxis=True, yAxisLabel="" if mini else "Contrib.\nscore",
        dataRange=cfg["importance"]["data-range"],
        mini=mini)

    logUtils.debug("Adding annotations.")
    usedNames = pu.addAnnotations(
        axAnnot=axAnnot, annotations=cfg["annotations"]["custom"],
        boxHeight=cfg["figure"]["annotation-height"],
        genomeStartX=genomeStartX, genomeEndX=genomeEndX,
        fontSize=cfg["figure"]["label-font-size"], mini=mini)
    # Now, add the profiles.
    logUtils.debug("Adding vertical plot.")
    pu.addVerticalProfilePlot(
        profile=cfg["predictions"]["values"][sliceStartY:sliceEndY],
        axProfile=axProfile,
        colors=cfg["predictions"]["color"][sliceStartY:sliceEndY],
        sequence=coords["sequence"][sliceStartY:sliceEndY],
        genomeWindowStart=coords["genome-window-start"] + sliceStartY,
        fontSizeTicks=cfg["figure"]["tick-font-size"],
        fontSizeAxLabel=cfg["figure"]["label-font-size"],
        dataRange=cfg["predictions"]["data-range"],
        mini=mini)
    if axLegend is not None:
        logUtils.debug("Adding legend.")
        pu.addLegend(usedNames=usedNames, axLegend=axLegend,
                     fontSize=cfg["figure"]["label-font-size"])
    logUtils.debug("Adding color bar.")
    pu.addCbar(pisaCax=pisaCax, axCbar=axCbar,
               fontSizeTicks=cfg["figure"]["tick-font-size"],
               fontSizeAxLabel=cfg["figure"]["label-font-size"],
               mini=mini)
    logUtils.debug("Finished making PISA plot.")
    # I return axes called colorbar and cbar. They are the same thing, but plotPisa
    # originally called it 'colorbar', while plotPisaGraph has always called it 'cbar'.
    # I return both to be more compatible. The documentation only calls it 'cbar'.
    # (Changed in 5.2.0)
    return {"axes": {"pisa": axPisa, "importance": axSeq, "predictions": axProfile,
                     "annotations": axAnnot, "colorbar": axCbar, "cbar": axCbar,
                     "legend": axLegend},
            "name-colors": cfg["annotations"].get("name-colors", {}),
            "genome-start": genomeStartX,
            "genome-end": genomeEndX,
            "config": cfg}


def plotModiscoPattern(pattern: motifUtils.Pattern,  # pylint: disable=too-many-statements
                       fig: matplotlib.figure.Figure, sortKey: NDArray | None = None) -> None:
    """Create a plot showing a pattern's seqlets and their match scores.

    :param sortKey: Either None (do not sort) or an array of shape (numSeqlets,)
        giving the order in which the seqlets should be displayed. See example below
        for common use cases.
    :type sortKey: None or ndarray
    :param pattern: The pattern to plot. This pattern must have already had its seqlets loaded.
    :param fig: The matplotlib figure upon which the plots should be drawn.

    Example::

        # Background ACGT frequency
        bgProbs = [0.291, 0.208, 0.208, 0.291]
        patSox2 = motifUtils.Pattern("pos_patterns", "pattern_0", "Sox2")
        with h5py.File(f"{WORK_DIR}/modisco/sox2_profile/modisco.h5", "r") as fp:
            patSox2.loadCwm(fp, 0.3, 0.3, bgProbs)
            patSox2.loadSeqlets(fp)
        fig = plt.figure(figsize=(5, 5))
        # Sort the seqlets by their contribution match.
        sortKey = [x.contribMatch for x in patSox2.seqlets]
        plotModiscoPattern(patSox2, fig, sortKey=sortKey)

    This produces the following plot:

    .. image:: ../../doc/presentations/modiscoPattern.png
        :width: 800
        :alt: Representative modisco pattern plot.

    """
    if sortKey is None:
        sortKey = np.arange(len(pattern.seqlets))
    sortOrder = np.argsort(sortKey)
    HIST_HEIGHT = 0.1  # pylint: disable=invalid-name
    PAD = 0.01  # pylint: disable=invalid-name
    axHmap = fig.add_axes((0.1, 0.1 + HIST_HEIGHT + PAD,
                           0.2 - PAD, 0.8 - HIST_HEIGHT))
    hmapAr = np.zeros((len(pattern.seqlets),
                       len(pattern.seqlets[0].sequence),
                       NUM_BASES), dtype=ONEHOT_T)
    for outIdx, seqletIdx in enumerate(sortOrder):
        hmapAr[outIdx] = utils.oneHotEncode(pattern.seqlets[seqletIdx].sequence)
    plotSequenceHeatmap(hmapAr, axHmap)
    axHmap.set_xticks([])
    hmapYticks = axHmap.get_yticks()[1:-1]
    axHmap.set_yticks(hmapYticks, [str(x) for x in hmapYticks], fontsize=FONT_SIZE_TICKS,
                      fontfamily=FONT_FAMILY)
    axLogo = fig.add_axes((0.1, 0.1, 0.2 - PAD, HIST_HEIGHT))
    cwm = pattern.cwm
    plotLogo(cwm, cwm.shape[0], axLogo, colors=bprcolors.dnaWong)
    axLogo.set_xlim(0, cwm.shape[0])
    cwmPos = np.zeros_like(cwm)
    cwmPos[cwm > 0] = cwm[cwm > 0]
    cwmNeg = np.zeros_like(cwm)
    cwmNeg[cwm < 0] = cwm[cwm < 0]
    sumPos = np.sum(cwmPos, axis=1)
    sumNeg = np.sum(cwmNeg, axis=1)

    axLogo.set_ylim(min(sumNeg), max(sumPos))
    axLogo.set_yticks([])
    axLogo.set_xticks([pattern.cwmTrimLeftPoint, pattern.cwmTrimRightPoint],
                      [str(pattern.cwmTrimLeftPoint), str(pattern.cwmTrimRightPoint)],
                      fontsize=FONT_SIZE_TICKS,
                      fontfamily=FONT_FAMILY)

    yvals = np.arange(len(pattern.seqlets), 0, -1)

    def plotStat(stat: np.ndarray, axPos: float, name: str, rightTicks: bool) -> AXES_T:
        """Plot a given statistic.

        :param stat: The values to be plotted.
        :param axPos: The position of the axes.
        :param name: The name of the statistic, for labeling the axis
        :param rightTicks: Should ticks be on the right side?
        :return: An Axes with the data plotted in it.
        """
        stat = np.array(stat)
        axCurStat = fig.add_axes((0.3 + axPos * 0.2, 0.1 + HIST_HEIGHT + PAD,
                                  0.2 - PAD, 0.8 - HIST_HEIGHT))
        statSort = stat[sortOrder]
        colorFix = parseSpec(bprcolors.defaultProfile)
        axCurStat.plot(statSort, yvals, ".", color=colorFix, alpha=0.5)
        statFilt = np.sort(statSort)
        colorFix = parseSpec({"tol": 2})
        axCurStat.plot(statFilt, yvals, "-", color=colorFix)
        axCurStat.set_ylim(0, len(pattern.seqlets))
        axCurStat.set_yticks([])
        axCurStat.set_title(name, fontdict={"fontsize": FONT_SIZE_LABELS,
                                            "fontfamily": FONT_FAMILY})
        axCurStat.set_xticks([])
        axCurStat.set_xlim(min(stat), max(stat))
        tickPoses = np.linspace(0, len(pattern.seqlets), 11, endpoint=True)
        tickLabels = np.arange(0, 110, 10)[::-1]
        axCurStat.tick_params(axis="y", labelleft=False, labelright=rightTicks,
                              left=False, right=rightTicks)
        axCurStat.set_yticks(tickPoses, tickLabels, fontsize=FONT_SIZE_TICKS,
                             fontfamily=FONT_FAMILY)
        axCurStat.grid()
        if rightTicks:
            axCurStat.yaxis.set_label_position("right")
            axCurStat.set_ylabel("Percentile", fontsize=FONT_SIZE_LABELS,
                                 fontfamily=FONT_FAMILY)
        axCurHist = fig.add_axes((0.3 + axPos * 0.2, 0.1, 0.2 - PAD, HIST_HEIGHT))
        hist = np.histogram(stat, bins=50)
        binMiddles = hist[1][:-1] + (hist[1][1] - hist[1][0]) / 2
        axCurHist.plot(binMiddles, hist[0])
        axCurHist.set_yticks([])
        axCurHist.set_xlim(min(stat), max(stat))
        xticks = axCurHist.get_xticks()
        xticks = xticks[1:-1]
        axCurHist.set_xticks(xticks, [str(x) for x in xticks], fontsize=FONT_SIZE_TICKS,
                             fontfamily=FONT_FAMILY)
        return axCurStat
    plotStat([x.seqMatch for x in pattern.seqlets], 0, "seqMatch", False)
    plotStat([x.contribMatch for x in pattern.seqlets], 1, "contribMatch", False)
    plotStat([x.contribMagnitude for x in pattern.seqlets], 2, "contribMag", True)


def plotSequenceHeatmap(hmap: ONEHOT_AR_T, ax: AXES_T, upsamplingFactor: int = 10) -> None:
    """Show a sequence heatmap from an array of one-hot encoded sequences.

    :param hmap: An array of sequences of shape (numSequences, length, NUM_BASES)
    :param ax: A matplotlib Axes object upon which the heatmap will be drawn.
    :param upsamplingFactor: How much should the x-axis be sharpened?
        If upsamplingFactor * hmap.shape[1] >> ax.width_in_pixels then you
        may get aliasing artifacts.
        If upsamplingFactor * hmap.shape[1] << ax.width_in_pixels then you
        will get blurry borders.

    """
    displayAr = np.zeros((hmap.shape[0], hmap.shape[1] * upsamplingFactor, 3),
                         dtype=PRED_T)
    for base, baseName in enumerate("ACGT"):
        hmapBase = np.array(hmap[:, :, base], dtype=PRED_T)
        ar = hmapBase
        for colorIdx in range(3):
            color = parseSpec(bprcolors.dnaWong[baseName])[colorIdx]
            for col in range(hmap.shape[1]):
                for colOffset in range(upsamplingFactor):
                    writeCol = col * upsamplingFactor + colOffset
                    displayAr[:, writeCol, colorIdx] += ar[:, col] * color
    ax.imshow(displayAr, aspect="auto", interpolation="antialiased",
              interpolation_stage="data")


def plotLogo(values: PRED_AR_T, width: float, ax: AXES_T,
             colors: DNA_COLOR_SPEC_T | list[DNA_COLOR_SPEC_T],
             spaceBetweenLetters: float = 0) -> None:
    """Plot an array of sequence data (like a pwm).

    :param values: An (N, NUM_BASES) array of sequence data. This could be, for example,
        a pwm or a one-hot encoded sequence.
    :param width: The width of the total logo, useful for aligning axis labels.
    :param ax: A matplotlib axes object on which the logo will be drawn.
    :param colors: The colors to use for shading the sequence. See below for details.
    :type colors: :class:`DNA_COLOR_SPEC_T<bpreveal.internal.constants.DNA_COLOR_SPEC_T>` |
        list[:class:`DNA_COLOR_SPEC_T<bpreveal.internal.constants.DNA_COLOR_SPEC_T>`]
    :param spaceBetweenLetters: How much should the letters be squished? This is
        given as a fraction of the total letter width. For example, to have
        a gap of 2 pixels between letters that are 10 pixels wide, set
        ``spaceBetweenLetters=0.2``.

    Colors, if provided, can have several meanings:
        1. Give a color for each base type by RGB value.
            In this case, colors will be a dict of tuples:
            ``{"A": (.8, .3, .2), "C": (.5, .3, .9), "G": (1., .4, .0), "T": (1., .7, 0.)}``
            This will make each instance of a particular base have the same color.
        2. Give a color for each base by color-spec.
            This would be something like:
            ``{"A": {"wong": 3}, "C": {"wong": 5}, "G": {"wong": 4}, "T": {"wong": 6}}``
            You can get the default BPReveal color map at
            :py:mod:`colors.dnaWong<bpreveal.colors.dnaWong>`.
        3. Give a list of colors for each base.
            This will be a list of length ``values.shape[0]`` and each entry
            should be a dictionary in either format
            1 or 2 above. This gives each base its own color palette, useful
            for shading bases by some profile.
    """
    # This is just a wrapper around a function that's in the internal library
    # to avoid circular imports.
    pu.plotLogo(values, width, ax, colors, spaceBetweenLetters)


def getCoordinateTicks(start: int, end: int, numTicks: int,
                       zeroOrigin: bool) -> tuple[list[float], list[str]]:
    """Given a start and end coordinate, return x-ticks that should be used for plotting.

    :param start: The genomic coordinate where your ticks start, inclusive.
    :param end: The genomic coordinate where your ticks end, *inclusive*.
    :param numTicks: The approximate number of ticks you want.
    :param zeroOrigin: The actual x coordinate of the ticks should start
        at zero, even though the labels start at ``start``. Otherwise,
        the ticks will be positioned at coordinate ``start`` to ``end``,
        and so your axes limits should actually correspond to genomic coordinates.
    :return: Two lists. The first is the x-coordinate of the ticks, and the second
        is the string labels that should be used at each tick.

    Given a start and end coordinate, return a list of ticks and tick labels that
    1. include exactly the start and stop coordinates
    2. Contain approximately numTicks positions and labels.
    3. Try to fall on easy multiples 1, 2, and 5 times powers of ten.
    4. Are formatted to reduce redundant label noise by omitting repeated initial digits.
    """
    # This is just a wrapper to avoid circular imports.
    return pu.getCoordinateTicks(start, end, numTicks, zeroOrigin)


def deleteTick(ax: AXES_T, which: Literal["x"] | Literal["y"] | Literal["both"],
               index: int) -> None:
    """Delete a particular offending tick from an Axes.

    :param ax: The axes that has the offending tick.
    :param which: Either ``"x"``, ``"y"``, or ``"both"``, indicating which axis
        has the problem.
    :param index: Which tick label index is the problem? This is a list index,
        and a negative number removes a particular label starting from the right.
        (Just like normal list indexing in Python.)


    This function helps when you have overlapping ticks, which happens annoyingly often
    since genomic coordinates tend to be huge numbers.

    Consider the following PISA plot:

    .. image:: ../../doc/presentations/pisaPlotMiniOverlappingLabels.png
        :width: 400
        :alt: Representative miniature PISA graph.

    As you can see, the label for ``'950`` is overlapping with the right-end label for
    ``180924977``. To delete this offending tick, I can use this function, like this::

        r = bprplots.plotPisa(plotConfig, fig)
        bprplots.deleteTick(r["axes"]["importance"], "x", -2)
        bprplots.deleteTick(r["axes"]["pisa"], "x", -2)

    Note that I had to remove the tick from both the importance and pisa axes.
    The deletion on the pisa axes causes the grid line to disappear.

    The result of this correction is the following:

    .. image:: ../../doc/presentations/pisaPlotMini.png
        :width: 400
        :alt: Representative miniature PISA graph.

    """
    def prune(ticks: Iterable, labels: Iterable) -> tuple[list, list[str]]:
        """Delete the offending tick and return new ones.

        :param ticks: The ticks that are ready to be drawn.
        :param labels: The labels for the ticks.
        :return: A tuple containing the new ticks and labels.
        """
        # We get numpy arrays in and want to do
        # concatenation with +, so we have to listify.
        ticks = list(ticks)
        labels = list(labels)
        if index < 0:
            # Slicing from the right/bottom
            breakPoint = len(ticks) + index
        else:
            breakPoint = index
        survivingTicks = ticks[:breakPoint] + ticks[breakPoint + 1:]
        survivingLabels = labels[:breakPoint] + labels[breakPoint + 1:]
        return survivingTicks, survivingLabels

    if which in {"x", "both"}:
        xTicks = ax.get_xticks()
        xLabels = ax.get_xticklabels()
        newXTicks, newXLabels = prune(xTicks, xLabels)
        ax.set_xticks(newXTicks)
        ax.set_xticklabels(newXLabels)

    if which in {"y", "both"}:
        yTicks = ax.get_yticks()
        yLabels = ax.get_yticklabels()
        newYTicks, newYLabels = prune(yTicks, yLabels)
        ax.set_yticks(newYTicks)
        ax.set_yticklabels(newYLabels)


# pylint: disable=missing-param-doc,missing-return-doc
def plotPisaWithFiles(pisaDats: str,  # pylint: disable=too-many-positional-arguments
                      cutMiddle: int, cutLengthX: int, cutLengthY: int, receptiveField: int,
                      genomeWindowStart: int,
                      genomeWindowChrom: str, genomeFastaFname: str, importanceBwFname: str,
                      motifScanBedFname: str, profileDats: str,
                      nameColors: dict[str, tuple[float, float, float]],
                      fig: matplotlib.figure.Figure, bbox: tuple[float, float, float, float],
                      colorSpan: float = 1.0, boxHeight: float = 0.1, fontsize: int = 5,
                      mini: bool = False) -> dict:
    """Create a new-style config dict from the old argument sea. Issues a warning if used."""
    del fig
    del receptiveField
    logUtils.error("This function has been replaced with the new plotting config style. "
                   "Use this config: ")
    cfg = {
        "pisa": {"h5-name": pisaDats},
        "coordinates": {
            "genome-fasta": genomeFastaFname,
            "midpoint-offset": cutMiddle,
            "input-slice-width": cutLengthX,
            "output-slice-width": cutLengthY,
            "genome-window-start": genomeWindowStart,
            "genome-window-chrom": genomeWindowChrom
        },
        "predictions": {"bigwig-name": profileDats},
        "importance": {"bigwig-name": importanceBwFname},
        "annotations": {"bed-name": motifScanBedFname,
                        "name-colors": nameColors},
        "figure": {
            "left": bbox[0],
            "bottom": bbox[1],
            "width": bbox[2],
            "height": bbox[3],
            "color-span": colorSpan,
            "annotation-height": boxHeight,
            "tick-font-size": fontsize,
            "label-font-size": fontsize,
            "miniature": mini
        }
    }
    logUtils.warning(str(cfg))
    return cfg

# pylint: enable=missing-param-doc,missing-return-doc



# Copyright 2022-2025 Charles McAnany. This file is part of BPReveal. BPReveal is free software: You can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version. BPReveal is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with BPReveal. If not, see <https://www.gnu.org/licenses/>.  # noqa
